src/views/trash-panel/trash-panel.tsx

```diff
@@ -0,0 +1,205 @@
+import { DataExplorer } from "~/views-components/data-explorer/data-explorer";
+import { DispatchProp, connect } from 'react-redux';
+import { DataColumns } from '~/components/data-table/data-table';
+import { RootState } from '~/store/store';
+import { DataTableFilterItem } from '~/components/data-table-filters/data-table-filters';
+import { SortDirection } from '~/components/data-table/data-column';
+import { ResourceKind, TrashableResource } from '~/models/resource';
+import { resourceLabel } from '~/common/labels';
+import { ArvadosTheme } from '~/common/custom-theme';
+import { RestoreFromTrashIcon, TrashIcon } from '~/components/icon/icon';
+import { TRASH_PANEL_ID } from "~/store/trash-panel/trash-panel-action";
+import { getProperty } from "~/store/properties/properties";
+import { PROJECT_PANEL_CURRENT_UUID } from "~/store/project-panel/project-panel-action";
+import { openContextMenu, resourceKindToContextMenuKind } from "~/store/context-menu/context-menu-actions";
+import { getResource, ResourcesState } from "~/store/resources/resources";
+import {
+    renderDate,
+    ResourceDeleteDate,
+    ResourceFileSize,
+    ResourceName,
+    ResourceTrashDate,
+    ResourceType
+} from "~/views-components/data-explorer/renderers";
+import { navigateTo } from "~/store/navigation/navigation-action";
+import { loadDetailsPanel } from "~/store/details-panel/details-panel-action";
+import { toggleCollectionTrashed, toggleProjectTrashed } from "~/store/trash/trash-actions";
+import { Dispatch } from "redux";
+
+type CssRules = "toolbar" | "button";
+
+const styles: StyleRulesCallback<CssRules> = (theme: ArvadosTheme) => ({
+    toolbar: {
+        paddingBottom: theme.spacing.unit * 3,
+        textAlign: "right"
+    },
+    button: {
+        marginLeft: theme.spacing.unit
+    },
+});
+
+export enum TrashPanelColumnNames {
+    NAME = "Name",
+    TYPE = "Type",
+    FILE_SIZE = "File size",
+    TRASHED_DATE = "Trashed date",
+    TO_BE_DELETED = "To be deleted"
+}
+
+export interface TrashPanelFilter extends DataTableFilterItem {
+    type: ResourceKind;
+}
+
+export const ResourceRestore =
+    connect((state: RootState, props: { uuid: string, dispatch?: Dispatch<any> }) => {
+        const resource = getResource<TrashableResource>(props.uuid)(state.resources);
+        return { resource, dispatch: props.dispatch };
+    })((props: { resource?: TrashableResource, dispatch?: Dispatch<any> }) =>
+        <IconButton onClick={() => {
+            if (props.resource && props.dispatch) {
+                const ctxRes = {
+                    name: '',
+                    uuid: props.resource.uuid,
+                    isTrashed: props.resource.isTrashed,
+                    ownerUuid: props.resource.ownerUuid
+                };
+
+                if (props.resource.kind === ResourceKind.PROJECT) {
+                    props.dispatch(toggleProjectTrashed(ctxRes));
```

Comment:
It could be done in the more component way:
```
const ResourceRestore = (props: {resource, onToggle}) => 
     <IconButton onClick={props.onToggle}>...</IconButton>

const mapStateToProps = (state, props) => ({resource: getResource(props.uuid)(state)});

const CollectionRestore = connect(mapDispatchToProps, dispatch => ({onToggle: dispatch(toggleCollection)}))(ResourceRestore)

const ProjectRestore = connect(mapDispatchToProps, dispatch => ({onToggle: dispatch(toggleProject)}}))(ResourceRestore)
```

Reply:
I tried, but that barely improves things as there must be somewhere switch on resource's kind :)